File: src/DataBase_Reader_VB/DatabaseManager.vb
Imports System.IO

Public MustInherit Class DatabaseManager
    Protected _dataBaseOpen As Boolean = False
    Protected _dataBasePath As String
    Protected _tableNames() As String
    Public ReadOnly Property DataBasePath As String
        Get
            Return _dataBasePath
        End Get
    End Property
    Public ReadOnly Property DataBaseOpen As Boolean
        Get
            Return _dataBaseOpen
        End Get
    End Property
    Public ReadOnly Property TableNames() As String()
        Get
            Return _tableNames
        End Get
    End Property

#Region "Events"
    Public Event EditsSaved(tableName As String, editsSaved As List(Of TableEdit))
    Public Sub OnEditsSaved(tableName As String, editsSaved As List(Of TableEdit))
        RaiseEvent editsSaved(tableName, editsSaved)
    End Sub
    Public Event PreviewEditsSaved(tableName As String, ByRef cancel As Boolean)
    Public Sub OnPreviewEditsSaved(tableName As String, ByRef cancel As Boolean)
        RaiseEvent PreviewEditsSaved(tableName, cancel)
    End Sub
    'Public Event RowsAdded(ByVal tableName As String, ByVal rowData As List(Of Object()))
    'Public Sub OnRowsAdded(ByVal tableName As String, ByVal rowData As List(Of Object()))
    '    RaiseEvent RowsAdded(tableName, rowData)
    'End Sub
    'Public Event RowsDeleted(ByVal tableName As String, ByVal rowIndices() As Int32)
    'Public Sub OnRowsDeleted(ByVal tableName As String, ByVal rowIndices() As Int32)
    '    RaiseEvent RowsDeleted(tableName, rowIndices)
    'End Sub
    'Public Event ColumnAdded(ByVal tableName As String, ByVal columnName As String, ByVal columnType As Type)
    'Public Sub OnColumnAdded(ByVal tableName As String, ByVal columnName As String, ByVal columnType As Type)
    '    RaiseEvent ColumnAdded(tableName, columnName, columnType)
    'End Sub
    'Public Event ColumnsDeleted(ByVal tableName As String, ByVal columnNames() As String)
    'Public Sub OnColumnsDeleted(ByVal tableName As String, ByVal columnNames() As String)
    '    RaiseEvent ColumnsDeleted(tableName, columnNames)
    'End Sub
#End Region



    Public MustOverride Sub Open()
    Public MustOverride Sub Close()
    Public MustOverride Function GetTableManager(ByVal tableName As String) As DataTableView
    Public MustOverride Function GetTableNames() As String()

    ''' <summary>
    ''' Get stored number of rows of a table.
    ''' </summary>
    ''' <param name="tableName">Table name.</param>
    ''' <returns></returns>
    Public MustOverride Function GetStoredNumberOfRows(tableName As String) As Long

    ''' <summary>
    ''' Get stored number of columns of a table.
    ''' </summary>
    ''' <param name="tableName">Table name.</param>
    ''' <returns></returns>
    Public MustOverride Function GetStoredNumberOfColumns(tableName As String) As Int32

    ''' <summary>
    ''' Determines if a type is numeric.  Nullable numeric types are considered numeric.
    ''' </summary>
    ''' <remarks>
    ''' Boolean is not considered numeric.
    ''' http://stackoverflow.com/questions/124411/using-net-how-can-i-determine-if-a-type-is-a-numeric-valuetype
    ''' </remarks>
    Public Shared Function IsNumericType(typeToTest As Type) As Boolean
        If typeToTest Is Nothing Then
            Return False
        End If

        Select Case Type.GetTypeCode(typeToTest)
            Case TypeCode.Byte, TypeCode.Decimal, TypeCode.Double, TypeCode.Int16, TypeCode.Int32, TypeCode.Int64,
             TypeCode.SByte, TypeCode.Single, TypeCode.UInt16, TypeCode.UInt32, TypeCode.UInt64
                Return True
            Case TypeCode.Object
                If typeToTest.IsGenericType AndAlso typeToTest.GetGenericTypeDefinition() = GetType(Nullable(Of )) Then
                    Return IsNumericType(Nullable.GetUnderlyingType(typeToTest))
                End If
                Return False
        End Select
        Return False
    End Function

    Public Shared Sub ConvertCsvToSqLite(inputFile As String, outputFile As String, outputtableName As String, hasHeaders As Boolean, dataLineStartIndex As Int32, fieldsEnclosedInQuotes As Boolean)
        'If Path.GetExtension(OutputFile).ToLower <> ".sqlite" Then Throw New Exception("Provided database file, " & System.IO.Path.GetFileName(OutputFile) & " is not a SQLite (.sqlite) file.")
        If Path.GetExtension(inputFile).ToLower <> ".csv" Then Throw New Exception("Provided database file, " & Path.GetFileName(outputFile) & " is not a comma separated (.csv) file.")
        If File.Exists(inputFile) = False Then Throw New Exception("Input database file, " & Path.GetFileName(outputFile) & " does not exist.")
        '
        Dim sqLiteEdit As New SQLiteManager(outputFile)
        sqLiteEdit.Open()
        Dim tableNames() As String = sqLiteEdit.GetTableNames
        For i As Int32 = 0 To tableNames.Count - 1
            If outputtableName = tableNames(i) Then Throw New Exception("Output table name already exists in the SQLite database file.")
        Next
        '
        Dim tempDataTable As New DataTable
        Dim columnHeaders As New List(Of String)
        Dim columnTypes As New List(Of Type)
        Using csvParser As New FileIO.TextFieldParser(inputFile) With {.TextFieldType = FileIO.FieldType.Delimited}
            csvParser.HasFieldsEnclosedInQuotes = fieldsEnclosedInQuotes
            csvParser.SetDelimiters(New String() {","})
            '
            Dim lineArray() As String = csvParser.ReadFields() 'csvRead.ReadLine().Split(",")
            If hasHeaders = True Then
                Dim headerCounter As Int32
                For Each header As String In lineArray
                    headerCounter = 1
                    Do Until tempDataTable.Columns.Contains(header) = False
                        header = header & headerCounter
                        headerCounter += 1
                    Loop
                    tempDataTable.Columns.Add(header, GetType(String))
                    columnHeaders.Add(header)
                    columnTypes.Add(GetType(String))
                Next
            Else
                For i As Int32 = 0 To lineArray.Count - 1
                    tempDataTable.Columns.Add("Column_" & i + 1, GetType(String))
                    columnHeaders.Add("Column_" & i + 1)
                    columnTypes.Add(GetType(String))
                Next
            End If
            Try
                sqLiteEdit.CreateTable(outputtableName, columnHeaders.ToArray, columnTypes.ToArray)
                Dim sqLiteTable As DataTableView = sqLiteEdit.GetTableManager(outputtableName)
                '
                For i As Integer = 0 To dataLineStartIndex - 1
                    lineArray = csvParser.ReadFields() 'csvRead.ReadLine()
                Next
                Dim lineCounter As Int32 = 0
                While Not csvParser.EndOfData
                    If lineArray.Count <> columnHeaders.Count Then Continue While
                    tempDataTable.Rows.Add(lineArray)
                    '
                    lineCounter += 1
                    If lineCounter > 500000 Then
                        sqLiteTable.AddRows(tempDataTable)
                        tempDataTable.Rows.Clear()
                        lineCounter = 0
                    End If
                    '
                    Try
                        lineArray = csvParser.ReadFields()
                    Catch ex As FileIO.MalformedLineException
                        'If the fields are enclosed in quotes then likely failed due to quotes within the quoted text
                        If fieldsEnclosedInQuotes Then lineArray = csvParser.ErrorLine.Substring(1, csvParser.ErrorLine.Length - 2).Split({Chr(34) & "," & Chr(34)}, StringSplitOptions.None)
                    End Try
                End While
                sqLiteTable.AddRows(tempDataTable)
                sqLiteTable.ApplyEdits()
                sqLiteEdit.Close()
            Catch ex As Exception
                Throw New Exception("Error occurred while writing sqlite table: " & vbCrLf & ex.Message)
            End Try
        End Using
    End Sub
End Class

File: src/DatabaseManager/DatabaseManager.cs
/*
* NOTICE:
* The U.S. Army Corps of Engineers, Risk Management Center (USACE-RMC) makes no guarantees about
* the results, or appropriateness of outputs, obtained from this software.
*
* LIST OF CONDITIONS:
* Redistribution and use in source and binary forms, with or without modification, are permitted
* provided that the following conditions are met:
* ● Redistributions of source code must retain the above notice, this list of conditions, and the
* following disclaimer.
* ● Redistributions in binary form must reproduce the above notice, this list of conditions, and
* the following disclaimer in the documentation and/or other materials provided with the distribution.
* ● The names of the U.S. Government, the U.S. Army Corps of Engineers, the Institute for Water
* Resources, or the Risk Management Center may not be used to endorse or promote products derived
* from this software without specific prior written permission. Nor may the names of its contributors
* be used to endorse or promote products derived from this software without specific prior
* written permission.
*
* DISCLAIMER:
* THIS SOFTWARE IS PROVIDED BY THE U.S. ARMY CORPS OF ENGINEERS RISK MANAGEMENT CENTER
* (USACE-RMC) "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
* THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
* DISCLAIMED. IN NO EVENT SHALL USACE-RMC BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
* SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
* PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
* INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
* LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
* THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

using System;
using System.Collections.Generic;
using System.Data;
using System.IO;
using System.Linq;

namespace DatabaseManager
{
    /// <summary>
    /// Abstract base class for managing database files, providing shared functionality such as event hooks and static helpers.
    /// </summary>
    public abstract class DatabaseManager
    {
        /// <summary>
        /// Indicates whether the database is currently open. 
        /// </summary>
        protected bool _dataBaseOpen = false;

        /// <summary>
        /// Path to the database file.
        /// </summary>
        protected string _dataBasePath;

        /// <summary>
        /// Array of table names in the database.
        /// </summary>
        protected string[] _tableNames;

        /// <summary>
        /// Gets the file path of the database.
        /// </summary>
        public string DataBasePath
        {
            get
            {
                return _dataBasePath;
            }
        }

        /// <summary>
        /// Gets whether the database is currently open.
        /// </summary>
        public bool DataBaseOpen
        {
            get
            {
                return _dataBaseOpen;
            }
        }

        /// <summary>
        /// Gets the list of table names contained in the database.
        /// </summary>
        public string[] TableNames
        {
            get
            {
                return _tableNames;
            }
        }

        #region Events

        /// <summary>
        /// Raised when edits are saved to a table.
        /// </summary>
        public event EditsSavedEventHandler EditsSaved;

        /// <summary>
        /// Delegate signature for the <see cref="EditsSaved"/> event.
        /// </summary>
        /// <param name="tableName">The name of the table whose edits were saved.</param>
        /// <param name="editsSaved">The list of edits that were saved to the database.</param>
        public delegate void EditsSavedEventHandler(string tableName, List<TableEdit> editsSaved);

        /// <summary>
        /// Raises the <see cref="EditsSaved"/> event after edits have been applied to the database.
        /// </summary>
        /// <param name="tableName">The name of the table whose edits were saved.</param>
        /// <param name="editsSaved">The list of edits that were saved to the database.</param>
        public void OnEditsSaved(string tableName, List<TableEdit> editsSaved)
        {
            EditsSaved?.Invoke(tableName, editsSaved);
        }

        /// <summary>
        /// Raised before edits are saved, allowing cancellation of the save operation.
        /// </summary>
        public event PreviewEditsSavedEventHandler PreviewEditsSaved;

        /// <summary>
        /// Delegate signature for the <see cref="PreviewEditsSaved"/> event.
        /// </summary>
        /// <param name="tableName">The name of the table being saved.</param>
        /// <param name="cancel">Reference parameter that can be set to true to cancel the save operation.</param>
        public delegate void PreviewEditsSavedEventHandler(string tableName, ref bool cancel);

        /// <summary>
        /// Raises the <see cref="PreviewEditsSaved"/> event before edits are applied to the database.
        /// </summary>
        /// <param name="tableName">The name of the table being saved.</param>
        /// <param name="cancel">Reference parameter that subscribers can set to true to cancel the save operation.</param>
        public void OnPreviewEditsSaved(string tableName, ref bool cancel)
        {
            PreviewEditsSaved?.Invoke(tableName, ref cancel);
        }

        #endregion

        /// <summary>
        /// Opens the database.
        /// </summary>
        public abstract void Open();

        /// <summary>
        /// Closes the database.
        /// </summary>
        public abstract void Close();

        /// <summary>
        /// Retrieves a table view manager for the specified table name.
        /// </summary>
        /// <param name="tableName">The name of the table to retrieve a manager for.</param>
        /// <returns>A <see cref="DataTableView"/> instance for managing the specified table.</returns>
        public abstract DataTableView GetTableManager(string tableName);

        /// <summary>
        /// Gets the list of table names in the database.
        /// </summary>
        /// <returns>An array of table names available in the database.</returns>
        public abstract string[] GetTableNames();

        /// <summary>
        /// Gets the stored number of rows in a table.
        /// </summary>
        /// <param name="tableName">The name of the table.</param>
        /// <returns>The number of rows stored in the table.</returns>
        public abstract long GetStoredNumberOfRows(string tableName);

        /// <summary>
        /// Gets the stored number of columns in a table.
        /// </summary>
        /// <param name="tableName">The name of the table.</param>
        /// <returns>The number of columns stored in the table.</returns>
        public abstract int GetStoredNumberOfColumns(string tableName);

        /// <summary>
        /// Determines if a type is numeric. Nullable numeric types are considered numeric.
        /// </summary>
        /// <param name="typeToTest">The <see cref="Type"/> to evaluate.</param>
        /// <returns><c>true</c> if the type is a numeric type; otherwise, <c>false</c>.</returns>
        /// <remarks>
        /// Boolean is not considered numeric.
        /// http://stackoverflow.com/questions/124411/using-net-how-can-i-determine-if-a-type-is-a-numeric-valuetype
        /// </remarks>
        public static bool IsNumericType(Type typeToTest)
        {
            if (typeToTest is null)
            {
                return false;
            }

            switch (Type.GetTypeCode(typeToTest))
            {
                case TypeCode.Byte:
                case TypeCode.Decimal:
                case TypeCode.Double:
                case TypeCode.Int16:
                case TypeCode.Int32:
                case TypeCode.Int64:
                case TypeCode.SByte:
                case TypeCode.Single:
                case TypeCode.UInt16:
                case TypeCode.UInt32:
                case TypeCode.UInt64:
                    {
                        return true;
                    }
                case TypeCode.Object:
                    {
                        if (typeToTest.IsGenericType && typeToTest.GetGenericTypeDefinition() == typeof(Nullable<>))
                        {
                            return IsNumericType(Nullable.GetUnderlyingType(typeToTest));
                        }
                        return false;
                    }
            }
            return false;
        }

        /// <summary>
        /// Converts a CSV file to an SQLite table.
        /// </summary>
        /// <param name="inputFile">The path to the input CSV file.</param>
        /// <param name="outputFile">The path to the output SQLite database file.</param>
        /// <param name="outputtableName">The name of the table to create in the SQLite database.</param>
        /// <param name="hasHeaders">Whether the CSV includes header rows.</param>
        /// <param name="dataLineStartIndex">The index at which data rows begin.</param>
        /// <param name="fieldsEnclosedInQuotes">Whether fields in the CSV are enclosed in quotes.</param>
        /// <exception cref="Exception">Thrown when file format is invalid or table already exists.</exception>
        public static void ConvertCsvToSqLite(string inputFile, string outputFile, string outputtableName, bool hasHeaders, int dataLineStartIndex, bool fieldsEnclosedInQuotes)
        {
            ArgumentException.ThrowIfNullOrWhiteSpace(inputFile);
            ArgumentException.ThrowIfNullOrWhiteSpace(outputFile);
            ArgumentException.ThrowIfNullOrWhiteSpace(outputtableName);

            if (Path.GetExtension(inputFile).ToLower() != ".csv")
            {
                throw new Exception("Provided input file, " + Path.GetFileName(inputFile) + " is not a comma separated (.csv) file.");
            }
            if (File.Exists(inputFile) == false)
            {
                throw new Exception("Input file, " + Path.GetFileName(inputFile) + " does not exist.");
            }

            var sqLiteEdit = new SQLiteManager(outputFile);
            try
            {
                sqLiteEdit.Open();
                string[] tableNames = sqLiteEdit.GetTableNames();
                for (int i = 0; i < tableNames.Length; i++)
                {
                    if ((outputtableName ?? "") == (tableNames[i] ?? ""))
                    {
                        throw new Exception("Output table name already exists in the SQLite database file.");
                    }
                }

                var tempDataTable = new DataTable();
                var columnHeaders = new List<string>();
                var columnTypes = new List<Type>();
                using (var csvParser = new Microsoft.VisualBasic.FileIO.TextFieldParser(inputFile) { TextFieldType = Microsoft.VisualBasic.FileIO.FieldType.Delimited })
                {
                    csvParser.HasFieldsEnclosedInQuotes = fieldsEnclosedInQuotes;
                    csvParser.SetDelimiters(new string[] { "," });

                    string[] lineArray = csvParser.ReadFields();
                    if (hasHeaders == true)
                    {
                        int headerCounter;
                        for (int i = 0; i < lineArray.Length; i++)
                        {
                            string header = lineArray[i];
                            headerCounter = 1;
                            while (tempDataTable.Columns.Contains(header) != false)
                            {
                                header = header + headerCounter;
                                headerCounter += 1;
                            }
                            tempDataTable.Columns.Add(header, typeof(string));
                            columnHeaders.Add(header);
                            columnTypes.Add(typeof(string));
                        }
                    }
                    else
                    {
                        for (int i = 0; i < lineArray.Length; i++)
                        {
                            tempDataTable.Columns.Add("Column_" + (i + 1), typeof(string));
                            columnHeaders.Add("Column_" + (i + 1));
                            columnTypes.Add(typeof(string));
                        }
                    }

                    sqLiteEdit.CreateTable(outputtableName, columnHeaders.ToArray(), columnTypes.ToArray());
                    var sqLiteTable = sqLiteEdit.GetTableManager(outputtableName);

                    for (int i = 0; i < dataLineStartIndex; i++)
                    {
                        lineArray = csvParser.ReadFields();
                    }
                    int lineCounter = 0;
                    while (!csvParser.EndOfData)
                    {
                        if (lineArray.Length != columnHeaders.Count) { continue; }
                        tempDataTable.Rows.Add(lineArray);

                        lineCounter += 1;
                        if (lineCounter > 500000)
                        {
                            sqLiteTable.AddRows(tempDataTable);
                            tempDataTable.Rows.Clear();
                            lineCounter = 0;
                        }

                        try
                        {
                            lineArray = csvParser.ReadFields();
                        }
                        catch (Microsoft.VisualBasic.FileIO.MalformedLineException)
                        {
                            // If the fields are enclosed in quotes then likely failed due to quotes within the quoted text
                            if (fieldsEnclosedInQuotes && !string.IsNullOrEmpty(csvParser.ErrorLine) && csvParser.ErrorLine.Length >= 2)
                            {
                                lineArray = csvParser.ErrorLine.Substring(1, csvParser.ErrorLine.Length - 2).Split(new[] { '"' + "," + '"' }, StringSplitOptions.None);
                            }
                        }
                    }
                    sqLiteTable.AddRows(tempDataTable);
                    sqLiteTable.ApplyEdits();
                }
            }
            catch (Exception ex)
            {
                throw new Exception("Error occurred while writing sqlite table: " + Environment.NewLine + ex.Message, ex);
            }
            finally
            {
                sqLiteEdit.Close();
            }
        }
    }
}